Ingester: allow only POST method on `/ingester/shutdown`

kevinmingtarja opened the pull request:
#### What this PR does
Previously, we allow both POST and GET `/ingester/shutdown`, which makes it too easy to accidentally trigger, i.e. internal scanners that traverse links can trigger shutdown of ingester just by visiting the link.

This PR makes that less likely to happen by only allowing POST requests on `/ingester/shutdown`. At the same time, if users still want to opt in to the existing behavior, we allow that by adding a new `-ingester.get-request-for-shutdown-enabled` flag (defaults to false). But this flag is also temporary and will be removed after two Mimir releases (2.15).

#### Which issue(s) this PR fixes or relates to

Fixes #7623

#### Checklist

- [x] Tests updated.
- [x] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/mimir?pullRequest=7707) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/mimir?pullRequest=7707) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/mimir?pullRequest=7707) it.</sub>

jhesketh reviewed (changes requested):
Thank you for your contribution!

There are a few tools and docs that will need updating to use POST:

https://github.com/grafana/mimir/blob/main/tools/migrate-ingester-statefulsets.sh#L46

https://github.com/grafana/mimir/blob/main/docs/sources/mimir/manage/run-production-environment/scaling-out.md?plain=1 (where it says invoke it should probably say POST).

https://github.com/grafana/mimir/blob/main/docs/sources/mimir/references/http-api/index.md?plain=1#L51 (and L412)

It would also be good to see a test to make sure GET is disabled unless the config flag is set.

pstibrany commented on pkg/api/api.go:
```diff
@@ -53,6 +53,9 @@ type Config struct {
 	EnableOtelMetadataStorage bool `yaml:"enable_otel_metadata_translation" category:"deprecated"`
 
+	// TODO: Remove option in Mimir 2.14.
+	AllowGETRequestForIngesterShutdown bool `yaml:"allow_get_request_for_ingester_shutdown" category:"deprecated"`
```
We use "-enabled" suffix for boolean flag. Please rename this field and flag to follow this convention.

pstibrany commented on CHANGELOG.md:
```diff
@@ -179,6 +179,7 @@
 * [BUGFIX] Fix metadata API using wrong JSON field names. #7475
 * [BUGFIX] Ruler: fix native histogram recording rule result corruption. #7552
 * [BUGFIX] Querier: fix HTTP status code translations for remote read requests. Previously, remote-read had conflicting behaviours: when returning samples all internal errors were translated to HTTP 400; when returning chunks all internal errors were translated to HTTP 500. #7487
+* [CHANGE] Ingester: allow only POST method on `/ingester/shutdown`, as previously it was too easy to accidentally trigger through GET requests. At the same time, add an option to keep the existing behavior by introducing an `-ingester.allow-get-request-for-shutdown` flag. This flag will be removed in Mimir 2.14. #7707
```
`CHANGE` is correct, as default changes. We keep CHANGE entries on the top of changelog for the release.

pstibrany commented on pkg/api/api.go:
```diff
@@ -296,7 +300,10 @@ func (a *API) RegisterIngester(i Ingester) {
 	a.RegisterRoute("/ingester/prepare-shutdown", http.HandlerFunc(i.PrepareShutdownHandler), false, true, "GET", "POST", "DELETE")
 	a.RegisterRoute("/ingester/prepare-partition-downscale", http.HandlerFunc(i.PreparePartitionDownscaleHandler), false, true, "GET", "POST", "DELETE")
-	a.RegisterRoute("/ingester/shutdown", http.HandlerFunc(i.ShutdownHandler), false, true, "GET", "POST")
+	a.RegisterRoute("/ingester/shutdown", http.HandlerFunc(i.ShutdownHandler), false, true, "POST")
+	if a.cfg.AllowGETRequestForIngesterShutdown {
+		a.RegisterDeprecatedRoute("/ingester/shutdown", http.HandlerFunc(i.ShutdownHandler), false, true, "GET")
+	}
```
Can we add test for this logic into `api_test.go`?

kevinmingtarja replied:
Yep, I've just added a test for this!

pstibrany commented on pkg/api/api.go:
```diff
@@ -71,6 +74,7 @@ type Config struct {
 func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 	f.BoolVar(&cfg.SkipLabelNameValidationHeader, "api.skip-label-name-validation-header-enabled", false, "Allows to skip label name validation via X-Mimir-SkipLabelNameValidation header on the http write path. Use with caution as it breaks PromQL. Allowing this for external clients allows any client to send invalid label names. After enabling it, requests with a specific HTTP header set to true wil ...
 	f.BoolVar(&cfg.EnableOtelMetadataStorage, "distributor.enable-otlp-metadata-storage", true, "If true, store metadata when ingesting metrics via OTLP. This makes metric descriptions and types available for metrics ingested via OTLP.")
+	f.BoolVar(&cfg.AllowGETRequestForIngesterShutdown, "ingester.allow-get-request-for-shutdown", false, "If true, allow GET requests to the /ingester/shutdown endpoint to trigger an ingester shutdown. This is a potentially dangerous operation and should only be enabled consciously.")
```
```suggestion
	f.BoolVar(&cfg.AllowGETRequestForIngesterShutdown, "ingester.allow-get-request-for-shutdown", false, "Enable GET requests to the /ingester/shutdown endpoint to trigger an ingester shutdown. This is a potentially dangerous operation and should only be enabled consciously.")
```

pstibrany reviewed (commented):
Thank you for working on this!

kevinmingtarja commented:
Hi @jhesketh, @pstibrany, thank you as well for the comments! 

I have updated the docs/tools that I missed, and addressed the comments in a follow-up commit. Please let me know if this is okay.

pstibrany commented on CHANGELOG.md:
```diff
@@ -125,6 +125,7 @@
 * [CHANGE] Distributor: the metric `cortex_distributor_sample_delay_seconds` has been deprecated and will be removed in Mimir 2.14. #7516
 * [CHANGE] Query-frontend: The deprecated YAML setting `frontend.cache_unaligned_requests` has been moved to `limits.cache_unaligned_requests`. #7519
 * [CHANGE] Querier: the CLI flag `-querier.minimize-ingester-requests` has been moved from "experimental" to "advanced". #7638
+* [CHANGE] Ingester: allow only POST method on `/ingester/shutdown`, as previously it was too easy to accidentally trigger through GET requests. At the same time, add an option to keep the existing behavior by introducing an `-ingester.get-request-for-shutdown-enabled` flag. This flag will be removed in Mimir 2.14. #7707
```
Since Mimir 2.12 was just released, and flag will be deprecated since Mimir 2.13, we can remove it only in Mimir 2.15.

pstibrany commented on pkg/api/api.go:
```diff
@@ -53,6 +53,9 @@ type Config struct {
 	// TODO: Remove option in Mimir 2.14.
 	EnableOtelMetadataStorage bool `yaml:"enable_otel_metadata_translation" category:"deprecated"`
 
+	// TODO: Remove option in Mimir 2.14.
```
```suggestion
	// TODO: Remove option in Mimir 2.15.
```

pstibrany commented on pkg/api/api_test.go:
```diff
@@ -206,6 +274,10 @@ func getServerConfig(t *testing.T) server.Config {
 
 		GRPCServerMaxRecvMsgSize: 1024,
+
+		// to avoid panic due to
+		// "duplicate metrics collector registration attempted"
+		Registerer: &MockRegisterer{},
```
We can simply create new registry. Some duplicate metrics error can point to a real problem, and it's good to catch those.

```suggestion
		// Use new registerer to avoid using default one and panic due to duplicate metrics.
		Registerer: prometheus.NewPedanticRegistry(),
```

We can then remove `MockRegisterer` type.

pstibrany reviewed (commented):
Thank you very much for working on this! I've left few minor comments.

jdbaldry reviewed (commented):
docs changes LGTM

pstibrany commented on pkg/api/api.go:
```diff
@@ -71,6 +74,7 @@ type Config struct {
 func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 	f.BoolVar(&cfg.SkipLabelNameValidationHeader, "api.skip-label-name-validation-header-enabled", false, "Allows to skip label name validation via X-Mimir-SkipLabelNameValidation header on the http write path. Use with caution as it breaks PromQL. Allowing this for external clients allows any client to send invalid label names. After enabling it, requests with a specific HTTP header set to true wil ...
 	f.BoolVar(&cfg.EnableOtelMetadataStorage, "distributor.enable-otlp-metadata-storage", true, "If true, store metadata when ingesting metrics via OTLP. This makes metric descriptions and types available for metrics ingested via OTLP.")
+	f.BoolVar(&cfg.GETRequestForIngesterShutdown, "ingester.get-request-for-shutdown-enabled", false, "Enable GET requests to the /ingester/shutdown endpoint to trigger an ingester shutdown. This is a potentially dangerous operation and should only be enabled consciously.")
```
I'm sorry, I've just noticed that CLI flag name (`-ingester.get-request-for-shutdown-enabled`) and yaml field name (`get_request_for_ingester_shutdown_enabled`) are not consistent. We should fix that, and also add `Enabled` suffix to field name in the struct.

```suggestion
	f.BoolVar(&cfg.GETRequestForIngesterShutdownEnabled, "ingester.get-request-for-ingester-shutdown-enabled", false, "Enable GET requests to the /ingester/shutdown endpoint to trigger an ingester shutdown. This is a potentially dangerous operation and should only be enabled consciously.")
```

Let's also fix CLI flag name in the changelog.

kevinmingtarja replied:
Ah no worries my bad, let me fix that real quick!

jhesketh reviewed (approved):
Thank you for the improvements!

jhesketh commented on cmd/mimir/help-all.txt.tmpl:
```diff
@@ -1249,6 +1249,8 @@ Usage of ./cmd/mimir/mimir:
   -flusher.exit-after-flush
     	Stop after flush has finished. If false, process will keep running, doing nothing. (default true)
+  -get-request-for-ingester-shutdown-enabled
+    	Enable GET requests to the /ingester/shutdown endpoint to trigger an ingester shutdown. This is a potentially dangerous operation and should only be enabled consciously. (default false)
```
```suggestion
    	[deprecated] Enable GET requests to the /ingester/shutdown endpoint to trigger an ingester shutdown. This is a potentially dangerous operation and should only be enabled consciously. (default false)
```

jhesketh reviewed (commented):
I think this will fix the unit test

jhesketh commented on pkg/api/api.go:
```diff
@@ -71,6 +74,7 @@ type Config struct {
 func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 	f.BoolVar(&cfg.SkipLabelNameValidationHeader, "api.skip-label-name-validation-header-enabled", false, "Allows to skip label name validation via X-Mimir-SkipLabelNameValidation header on the http write path. Use with caution as it breaks PromQL. Allowing this for external clients allows any client to send invalid label names. After enabling it, requests with a specific HTTP header set to true wil ...
 	f.BoolVar(&cfg.EnableOtelMetadataStorage, "distributor.enable-otlp-metadata-storage", true, "If true, store metadata when ingesting metrics via OTLP. This makes metric descriptions and types available for metrics ingested via OTLP.")
+	f.BoolVar(&cfg.GETRequestForIngesterShutdownEnabled, "get-request-for-ingester-shutdown-enabled", false, "Enable GET requests to the /ingester/shutdown endpoint to trigger an ingester shutdown. This is a potentially dangerous operation and should only be enabled consciously.")
```
On re-review, this should probably be:
```suggestion
	f.BoolVar(&cfg.GETRequestForIngesterShutdownEnabled, "ingester.get-request-for-shutdown-enabled", false, "Enable GET requests to the /ingester/shutdown endpoint to trigger an ingester shutdown. This is a potentially dangerous operation and should only be enabled consciously.")
```
And the corresponding changes to the docs.

kevinmingtarja replied:
Got it, should the `yaml` field name stay the same though?

kevinmingtarja replied:
Actually, initially I did as you suggested, but then Peter suggested to make it consistent so I changed it: https://github.com/grafana/mimir/pull/7707#discussion_r1551589188

jhesketh replied:
Oh right, I missed that conversation sorry. Looking at the `distributor.enable-otlp-metadata-storage` flag (just above) it exists as both `distributor.enable-otlp-metadata-storage` as the flag and `enable_otel_metadata_translation` as the yaml. @pstibrany Just want to double check your opinion here please.

pstibrany replied:
I'm sorry for confusion here. The problem here comes from the fact that config option is defined in `api` config block in YAML, but we're trying to use `ingester.` prefix for CLI flag. At the same time, we're tring to make CLI flag and yaml field name consistent (this was goal of my previous comment, without realizing it's causing more confusion).

I think we should settle on:
* `-api.get-request-for-ingester-shutdown-enabled` for CLI flag
* `get_request_for_ingester_shutdown_enabled` as YAML field name (it's already under `api` section)

If we want to use `-ingester.` prefix for CLI flag, then we should put this into ingester config struct.

Unfortunately I think `enable_otel_metadata_translation` was a mistake, as it doesn't follow our practices. ("enabled" suffix, consistent CLI and YAML names)

pstibrany replied:
@jhesketh thanks for catching this. My [suggestion](https://github.com/grafana/mimir/pull/7707#discussion_r1551589188) was not to remove `-ingester.` prefix (only to add `ingester` in the middle of the name), and I missed that this is what happened. 🤦

kevinmingtarja replied:
Ah sorry for the misunderstanding too. So should we keep going with the `-ingester.` prefix or `-api.` instead? As you mentioned, I think the latter makes more sense as it's already under the `api` section in the yaml.

pstibrany replied:
Hi Kevin, with Josh we agreed to go with `-api.get-request-for-ingester-shutdown-enabled` CLI flag and `get_request_for_ingester_shutdown_enabled` YAML field name. Thank you very much for your patience with us!

kevinmingtarja replied:
Hi Peter, no worries! Let me quickly do that

jhesketh reviewed (approved):
LGTM! Thank you for persisting with our miss on the naming :-). Will leave it for Peter to do final look+merge.

kevinmingtarja commented:
Hang on did i miss something, somehow one of the test fails with this, even though `(default false)` is there in `help-all.txt.tmpl`.

```
Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -282,3 +282,3 @@
            	            	   -api.get-request-for-ingester-shutdown-enabled
            	            	-    	[deprecated] Enable GET requests to the /ingester/shutdown endpoint to trigger an ingester shutdown. This is a potentially dangerous operation and should only be enabled consciously. (default false)
            	            	+    	[deprecated] Enable GET requests to the /ingester/shutdown endpoint to trigger an ingester shutdown. This is a potentially dangerous operation and should only be enabled consciously.
            	            	   -api.skip-label-name-validation-header-enabled
            	Test:       	TestHelp/all
            	Messages:   	./cmd/mimir/mimir -help-all output changed; try `make reference-help`
```

pstibrany reviewed (approved):
Thank you very much for your contribution and patience with us!

pstibrany commented:
> Hang on did i miss something, somehow one of the test fails with this, even though `(default false)` is there in `help-all.txt.tmpl`.

I've regenerated help (using `make reference-help`) and push that change to resolve this problem.

kevinmingtarja commented:
No worries, thanks as well for the review Peter and Joshua!
